rubygems/index.js

```diff
@@ -0,0 +1,230 @@
+const got = require('got')
+const compareVersions = require('compare-versions')
+const limit = require('call-limit')
+
+class RubyGemsDependencyResolver {
+  constructor({ log }) {
+    this.log = log
+    this.got = limit.promise(got, 30)
+  }
+
+  // a regex-type string list that represents the search pattern
+  // for this language/registry's manifest files
+  getManifestPatterns() {
+    return ["Gemfile"]
+  }
+
+  // Def: given a raw file (i.e. the bytes of Gemfile), return a list ([]) of
+  //  dependencies that are listed in the manifest file. The format of each entry in the list
+  //  should be consumable (one at a time) by getSpec.
+  extractDependenciesFromManifest({ manifest }) {
+
+    // TODO: If we detect that the source of the manifest ISN"T rubygems.org, should we even parse the deps??
+
+    // TODO: doesn't support `when '1.9.1'; gem 'ruby-debug-base19', '0.11.23'` syntax
+
+    if (!manifest) return []
+    const extract = manifest.split("\n").reduce((acc, line) => {
+      /**
+       * trim off leading and trailing whitespace
+       * and lowercase the requirement to avoid duplication
+       */
+      line = line.trim().toLowerCase()
+      /**
+       * skip all blank lines
+       */
+      if (line === "") return acc
+      /**
+       * skip all lines that are comments
+       */
+      if (line.startsWith("#")) return acc
+      /**
+       * Add all lines that start with "gem" since we're including deps from every group
+       */
+      if (line.startsWith("gem")) {
+        // Split on # and remove the comments at the end of any lines
+        const lineWithoutComment = line.split('#')[0]
+        return acc.concat(lineWithoutComment)
+      }
+      return acc
+    }, [])
+    return extract
+  }
+
+  /**
+   * - input: a single entry from extractDependenciesFromManifest, input format is of 
+   * "gem <name of gem>, "version (optional)", specifiers (optional)
+   */
+  getSpec(pkg) {
+    if (typeof pkg === "object") return pkg
+
+    // Version seems to always come after the gem name, if it is specified. 
+    const pkgParts = pkg.split(",")
+    // remove gem, trim off white space, remove quotes, now we have the name
+    const name = pkgParts[0].replace('gem', '').trim().replace(/'/g, '').replace(/"/g, '')
+
+    // If there's nothing after the name, then return latest as version
+    if (!pkgParts[1]) {
+      return {
+        name,
+        operator: '==',
+        versionSpec: 'latest',
+        toString: () => `${name}@`
+      }
+    }
+
+    const re = /^('|")(==|=|>=|>|<=|~>)?\s*([0-9.]+)('|")$/
+    const match = pkgParts[1].trim().match(re)
+    if (!match) throw new Error(`unparseable pkg dependency ${pkgParts[1]}`)
+    const operator = match[2] || '='
```

Comment:
what is the difference between this fallback `=` and `==` on L70?

Reply:
two cases:

gem 'rubocop'
gem 'rubocop', '1.0.0'

in first case we want to set operator to == (or =, doesn't matter) and version to "latest"
in second case we want operator to be = or == and version to be whatever is set